theseus/src/api/profile.rs

```diff
@@ -0,0 +1,196 @@
+//! Theseus profile management interface
+
+pub use crate::{
+    state::{JavaSettings, Profile},
+    State,
+};
+use daedalus as d;
+use std::{future::Future, path::Path};
+use tokio::process::{Child, Command};
+
+/// Add a profile to the in-memory state
+pub async fn add(profile: Profile) -> crate::Result<()> {
+    let state = State::get().await?;
+    let mut profiles = state.profiles.write().await;
+    profiles.insert(profile)?;
+
+    Ok(())
+}
+
+/// Add a path as a profile in-memory
+pub async fn add_path(path: &Path) -> crate::Result<()> {
+    let state = State::get().await?;
+    let mut profiles = state.profiles.write().await;
+    profiles.insert_from(path).await?;
+
+    Ok(())
+}
+
+/// Remove a profile
+pub async fn remove(path: &Path) -> crate::Result<()> {
+    let state = State::get().await?;
+    let mut profiles = state.profiles.write().await;
+    profiles.remove(path)?;
+
+    Ok(())
+}
+
+/// Get a profile by path
+pub async fn get(path: &Path) -> crate::Result<Option<Profile>> {
+    let state = State::get().await?;
+    let profiles = state.profiles.read().await;
+
+    Ok(profiles.0.get(path).cloned())
+}
+
+/// Check if a profile is already managed by Theseus
+pub async fn is_managed(profile: &Path) -> bool {
+    let state = State::get().await.unwrap();
+    let profiles = state.profiles.read().await;
+    profiles.0.contains_key(profile)
+}
+
+/// Edit a profile using a given asynchronous closure
+pub async fn edit<Fut>(
+    path: &Path,
+    action: impl Fn(&mut Profile) -> Fut,
+) -> crate::Result<()>
+where
+    Fut: Future<Output = crate::Result<()>>,
+{
+    let state = State::get().await.unwrap();
+    let mut profiles = state.profiles.write().await;
+    if let Some(profile) = profiles.0.get_mut(path) {
+        action(profile).await
+    } else {
+        Err(crate::Error::OtherError(format!(
+            "Tried to edit a nonexistent profile at path {}!",
+            path.display()
+        )))
+    }
+}
+
+/// Run Minecraft using a profile
+pub async fn run(
+    path: &Path,
+    credentials: &crate::launcher::Credentials,
+) -> crate::Result<Child> {
```

Comment:
Might be out of scope, but should we have some sort of input mut buffer for the instance log for displaying?

Or we could do it later when that needs to be done

Reply:
Probably should wait on that